lib/property_table/persist.ex

```diff
@@ -0,0 +1,229 @@
+        error
+    end
+  rescue
+    e ->
+      Logger.error("Failed to restore table from disk: #{e}")
+      {:error, :failed_to_restore}
+  end
+
+  @spec save_snapshot(reference() | atom(), Keyword.t()) ::
+          {:ok, String.t()} | :error | no_return()
+  def save_snapshot(table, options) do
+    options = take_options(options)
+    persist_to_disk(table, options)
+
+    snapshot_id = :crypto.strong_rand_bytes(8) |> Base.encode16()
+
+    full_snapshot_name = "#{snapshot_id}"
+    stable_path = get_path(:stable, options)
+    snapshot_path = get_path(:snapshot, options, full_snapshot_name)
+
+    # Move file to snapshot directory
+    Logger.debug("Creating table file snapshot: #{snapshot_path}")
+    File.copy!(stable_path, snapshot_path)
+
+    maybe_clean_old_snapshots(options)
+
+    {:ok, snapshot_id}
+  rescue
+    e ->
+      Logger.error("Failed to save snapshot to disk: #{e}")
+      :error
+  end
+
+  @spec restore_snapshot(Keyword.t(), String.t()) :: :ok | {:error, :enoent} | no_return()
+  def restore_snapshot(options, snapshot_id) do
+    options = take_options(options)
+    stable_path = get_path(:stable, options)
+    snapshots_path = get_path(:snapshot, options) |> Path.join("#{snapshot_id}")
+    found_snapshot = Path.wildcard(snapshots_path)
+
+    if length(found_snapshot) != 1 do
+      {:error, :enoent}
+    else
+      [to_restore] = found_snapshot
+
+      # Copy the found snapshot in place of the current stable db path
+      Logger.debug("Restoring table file snapshot: #{to_restore} => #{stable_path}")
+      File.copy!(to_restore, stable_path)
+
+      :ok
+    end
+  end
+
+  @spec get_snapshot_list(Keyword.t()) :: [{String.t(), tuple()}] | no_return()
+  def get_snapshot_list(options) do
+    options = take_options(options)
+    snapshot_path = get_path(:snapshot, options)
+
+    snapshots =
+      File.ls!(snapshot_path)
+      |> Enum.map(fn id ->
+        stat = Path.join(snapshot_path, [id]) |> File.stat!()
+        {id, stat.ctime}
+      end)
+      |> Enum.sort_by(fn {_id, ctime} -> ctime end)
+
+    snapshots
+  end
+
+  defp try_restore_tabfile(tabfile_path) do
+    Logger.debug("Attempting to load data from file: #{tabfile_path}")
+
+    case PersistFile.decode_file(tabfile_path) do
+      {:ok, data} ->
+        {:halt, {:ok, :erlang.binary_to_term(data)}}
+
+      {:error, err} ->
+        Logger.warn("Failed to load data from file #{tabfile_path} - #{inspect(err)}")
+        Logger.warn("Will try another backup file...")
+        {:cont, {:error, err}}
+    end
+  end
+
+  defp take_options(options) when is_list(options) do
+    @persist_options
+    |> Enum.map(fn {key_name, default_value} ->
+      {key_name, Keyword.get(options, key_name, default_value)}
+    end)
+  end
+
+  defp get_path(:stable, options) do
+    dir = Path.join(options[:data_directory], options[:table_name])
+    File.mkdir_p!(dir)
+    Path.join(options[:data_directory], [options[:table_name], "/", @data_stable_name])
```

Comment:
```suggestion
    Path.join([options[:data_directory], options[:table_name], @data_stable_name])
```

I might have personal preference against using `Path.t`'s with path separators in `Path.join/2`. But thinking about it, I actually don't know whether `Path.join/2` fixes up path separators that are passed into the second param. 

I changed this to a `Path.join/1` since I find it less ambiguous for joining >2 path elements. WDYT

Reply:
I'm on board with moving to `Path.join/1`